File: python/24.py
# Definition for singly-linked list.
# n = length of linked list
# O(n)
# 將所有的元素按照單雙存成兩個list
# 再反向存回去

class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        # margin testcase
        if not head:
            return None
        if not head.next:
            return head
        order_list_1 = []
        order_list_2 = []
        cur = head
        cur_idx = 0
        while cur:
            if cur_idx % 2 == 0:
                order_list_1.append(cur)
            else:
                order_list_2.append(cur)
            cur = cur.next
            cur_idx += 1
        min_len = min(len(order_list_1), len(order_list_2))
        ans_head = ans_cur = order_list_2[0]
        ans_cur.next = order_list_1[0]
        ans_cur = ans_cur.next
        for i in range(1, min_len):
            ans_cur.next = order_list_2[i]
            ans_cur = ans_cur.next
            ans_cur.next = order_list_1[i]
            ans_cur = ans_cur.next
        if len(order_list_1) == min_len+1:
            ans_cur.next = order_list_1[min_len]
            ans_cur = ans_cur.next
        ans_cur.next = None
        ans_cur = ans_head
        # while ans_cur:
        #     print(ans_cur.val)
        #     ans_cur = ans_cur.next
        return ans_head


if __name__ == "__main__":
    sol = Solution()
    head = ListNode(1)
    t = ListNode(2)
    tt = ListNode(3)
    ttt = ListNode(4)
    head.next = t
    t.next = tt
    tt.next = ttt
    sol.swapPairs(head)

File: c/24.c
#include <stdio.h>
#include <stdlib.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* swapPairs(struct ListNode* head){
	struct ListNode *L1, *L2, *p1, *p2, *cur, *tmp;
	int i = 0;
	if(!head){
		return NULL;
	}
	L1 = head;
	L2 = head->next;
	p1 = head;
	p2 = head->next;

	if(!L2){
		return head;
	}
	cur = L2->next;

	while(cur){
		// printf("cur: %d\n", cur->val);
		if(i%2 == 0){
			p1->next = cur;
			p1 = p1->next;
		}else{
			p2->next = cur;
			p2 = p2->next;
		}
		cur = cur->next;
		i++;
	}
	p1->next = NULL;
	p2->next = NULL;
	
	head = L2;
	p1 = L1;
	p2 = L2;
	while(p1 && p2){
		tmp = p2->next;
		p2->next = p1;
		p2 = tmp;

		tmp = p1->next;
		if(p2){
			p1->next = p2;
		}
		p1 = tmp;
	}
	return L2;
}


int main(){
	struct ListNode *n1, *n2, *n3, *n4;
	struct ListNode *cur, *ans;

	n1 = malloc(sizeof(struct ListNode* ));
	n2 = malloc(sizeof(struct ListNode* ));
	n3 = malloc(sizeof(struct ListNode* ));
	n4 = malloc(sizeof(struct ListNode* ));

	n1->val = 1;
	n2->val = 2;
	n3->val = 3;
	n4->val = 4;
	n1->next = NULL;
	n2->next = n3;
	n3->next = NULL;
	n4->next = NULL;
	ans = swapPairs(n1);
	cur = ans;
	while(cur){
		printf("%d\n", cur->val);
		cur = cur->next;
	}
	return 0;
}









